core/src/main/java/aima/core/search/basic/informed/RecursiveBestFirstSearch.java

```diff
@@ -52,27 +52,27 @@
 	@Override
 	public List<A> apply(Problem<A, S> problem) {
 		// return RBFS(problem, MAKE-NODE(problem.INITIAL-STATE), infinity)
-		return rbfs(problem, new SuccessorNode(nodeFactory.newRootNode(problem.initialState()), this::h),
+		return rbfs(problem, newRootNode(problem.initialState(), 0),
 				Double.POSITIVE_INFINITY).result();
 	}
 
 	// function RBFS(problem, node, f_limit) returns a solution, or failure and
 	// a new f-cost limit
 	public Result rbfs(Problem<A, S> problem, SuccessorNode node, double f_limit) {
 		// if problem.GOAL-TEST(node.STATE) then return SOLUTION(node)
-		if (searchController.isGoalState(node.n, problem)) {
-			return new Result(searchController.solution(node.n));
+		if (problem.isGoalState(node.state())) {

```

Comment:
You want to add a 'isGoalState()' helper function call here, that actually calls the searchController like before i.e.:

searchController.isGoalState(node.n, problem)

and not the problem directly as you have done. The reason for this is it lets someone override the behavior in the search controller, for e.g. if they want to visit more than 1 goal state. The reason for passing the Node to the searchController function - problem.isGoalState only gets to check the state and not necessarily the node information associated with it.